### PR TITLE
Chore: golang 1.21

### PR DESCRIPTION
**Pull request by shyukri**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

- Fixes https://github.com/argoproj/argo-events/issues/2742
- Bumps alpine base to 3.19 and ArgoWF client

This has also gone bigger than anticipated:
- bump golangci-lint; drop abandoned linters && add recommendations
- move k3d to 1.27 and fix CI
- update github action setup-go to v5
- ...

### CODE REVIEW

**Comment by whynowy** [comment]
Thanks @shyukri ! could you figure out why the lint is failing?

**Comment by shyukri** [comment]
Issues even after upgrading the linter:

```
# make lint

INFO [runner] linters took 26.82040325s with stages: goanalysis_metalinter: 26.7034705s
test/util/util.go:21:19: string `metadata.name=` has 3 occurrences, make it a constant (goconst)
	fieldSelector := "metadata.name=" + eventBusName
	                 ^
controllers/eventbus/installer/jetstream_test.go:74:18: string `eventbus-` has 6 occurrences, make it a constant (goconst)
	assert.Equal(t, "eventbus-"+testJetStreamEventBus.Name+"-js", n)
	                ^
common/string_keyed_map.go:7:2: `items` is unused (structcheck)
	items map[string]T
	^
common/string_keyed_map.go:8:2: `lock` is unused (structcheck)
	lock  *sync.RWMutex
	^
controllers/config.go:7:2: import 'github.com/fsnotify/fsnotify' is not allowed from list 'Main' (depguard)
	"github.com/fsnotify/fsnotify"
	^
controllers/config.go:8:2: import 'github.com/spf13/viper' is not allowed from list 'Main' (depguard)
	"github.com/spf13/viper"
	^
eventbus/jetstream/base/jetstream.go:8:2: import 'github.com/argoproj/argo-events/common' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common"
	^
eventbus/jetstream/base/jetstream.go:9:2: import 'github.com/argoproj/argo-events/eventbus/common' is not allowed from list 'Main' (depguard)
	eventbuscommon "github.com/argoproj/argo-events/eventbus/common"
	^
eventbus/jetstream/base/jetstream.go:10:2: import 'github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1' is not allowed from list 'Main' (depguard)
	eventbusv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1"
	^
eventbus/jetstream/base/jetstream.go:11:2: import 'github.com/nats-io/nats.go' is not allowed from list 'Main' (depguard)
	nats "github.com/nats-io/nats.go"
	^
eventbus/jetstream/base/jetstream.go:12:2: import 'github.com/spf13/viper' is not allowed from list 'Main' (depguard)
	"github.com/spf13/viper"
	^
eventbus/jetstream/base/jetstream_conn.go:6:2: import 'github.com/nats-io/nats.go' is not allowed from list 'Main' (depguard)
	nats "github.com/nats-io/nats.go"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:9:2: import 'github.com/argoproj/argo-events/eventbus/common' is not allowed from list 'Main' (depguard)
	eventbuscommon "github.com/argoproj/argo-events/eventbus/common"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:10:2: import 'github.com/argoproj/argo-events/eventbus/jetstream/base' is not allowed from list 'Main' (depguard)
	eventbusjetstreambase "github.com/argoproj/argo-events/eventbus/jetstream/base"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:11:2: import 'github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:12:2: import 'github.com/nats-io/nats.go' is not allowed from list 'Main' (depguard)
	nats "github.com/nats-io/nats.go"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:17:2: import 'github.com/argoproj/argo-events/common' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:18:2: import 'github.com/cloudevents/sdk-go/v2' is not allowed from list 'Main' (depguard)
	cloudevents "github.com/cloudevents/sdk-go/v2"
	^
eventbus/jetstream/sensor/sensor_jetstream.go:19:2: import 'github.com/mitchellh/hashstructure/v2' is not allowed from list 'Main' (depguard)
	hashstructure "github.com/mitchellh/hashstructure/v2"
	^
eventbus/jetstream/sensor/trigger_conn.go:12:2: import 'github.com/Knetic/govaluate' is not allowed from list 'Main' (depguard)
	"github.com/Knetic/govaluate"
	^
eventbus/jetstream/sensor/trigger_conn.go:13:2: import 'github.com/cloudevents/sdk-go/v2' is not allowed from list 'Main' (depguard)
	cloudevents "github.com/cloudevents/sdk-go/v2"
	^
eventbus/jetstream/sensor/trigger_conn.go:16:2: import 'github.com/argoproj/argo-events/eventbus/common' is not allowed from list 'Main' (depguard)
	eventbuscommon "github.com/argoproj/argo-events/eventbus/common"
	^
eventbus/jetstream/sensor/trigger_conn.go:17:2: import 'github.com/argoproj/argo-events/eventbus/jetstream/base' is not allowed from list 'Main' (depguard)
	jetstreambase "github.com/argoproj/argo-events/eventbus/jetstream/base"
	^
eventbus/kafka/eventsource/source_conn.go:6:2: import 'github.com/IBM/sarama' is not allowed from list 'Main' (depguard)
	"github.com/IBM/sarama"
	^
eventbus/kafka/eventsource/source_conn.go:8:2: import 'github.com/argoproj/argo-events/eventbus/kafka/base' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/eventbus/kafka/base"
	^
eventbus/kafka/eventsource/source_kafka.go:4:2: import 'github.com/IBM/sarama' is not allowed from list 'Main' (depguard)
	"github.com/IBM/sarama"
	^
eventbus/kafka/eventsource/source_kafka.go:6:2: import 'github.com/argoproj/argo-events/eventbus/kafka/base' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/eventbus/kafka/base"
	^
eventbus/kafka/eventsource/source_kafka.go:7:2: import 'github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1' is not allowed from list 'Main' (depguard)
	eventbusv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1"
	^
eventbus/kafka/sensor/kafka_handler.go:9:2: import 'github.com/IBM/sarama' is not allowed from list 'Main' (depguard)
	"github.com/IBM/sarama"
	^
eventbus/kafka/sensor/kafka_handler.go:10:2: import 'github.com/argoproj/argo-events/eventbus/kafka/base' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/eventbus/kafka/base"
	^
eventbus/kafka/sensor/kafka_sensor.go:12:2: import 'github.com/Knetic/govaluate' is not allowed from list 'Main' (depguard)
	"github.com/Knetic/govaluate"
	^
eventbus/kafka/sensor/kafka_sensor.go:15:2: import 'github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1' is not allowed from list 'Main' (depguard)
	eventbusv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1"
	^
eventbus/kafka/sensor/kafka_sensor.go:16:2: import 'github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1' is not allowed from list 'Main' (depguard)
	sensorv1alpha1 "github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
	^
eventbus/kafka/sensor/kafka_sensor.go:17:2: import 'github.com/cloudevents/sdk-go/v2' is not allowed from list 'Main' (depguard)
	cloudevents "github.com/cloudevents/sdk-go/v2"
	^
eventbus/kafka/sensor/trigger_conn.go:8:2: import 'github.com/Knetic/govaluate' is not allowed from list 'Main' (depguard)
	"github.com/Knetic/govaluate"
	^
eventbus/stan/base/stan.go:9:2: import 'github.com/nats-io/stan.go' is not allowed from list 'Main' (depguard)
	"github.com/nats-io/stan.go"
	^
eventbus/stan/base/stan_conn.go:7:2: import 'github.com/nats-io/stan.go' is not allowed from list 'Main' (depguard)
	"github.com/nats-io/stan.go"
	^
eventbus/stan/eventsource/source_conn.go:8:2: import 'github.com/argoproj/argo-events/eventbus/stan/base' is not allowed from list 'Main' (depguard)
	stanbase "github.com/argoproj/argo-events/eventbus/stan/base"
	^
eventbus/stan/eventsource/source_stan.go:5:2: import 'github.com/argoproj/argo-events/eventbus/stan/base' is not allowed from list 'Main' (depguard)
	stanbase "github.com/argoproj/argo-events/eventbus/stan/base"
	^
eventbus/stan/sensor/sensor_stan.go:9:2: import 'github.com/argoproj/argo-events/common' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common"
	^
eventbus/stan/sensor/sensor_stan.go:11:2: import 'github.com/argoproj/argo-events/eventbus/stan/base' is not allowed from list 'Main' (depguard)
	stanbase "github.com/argoproj/argo-events/eventbus/stan/base"
	^
eventbus/stan/sensor/trigger_conn.go:13:2: import 'github.com/gobwas/glob' is not allowed from list 'Main' (depguard)
	"github.com/gobwas/glob"
	^
eventbus/stan/sensor/trigger_conn.go:14:2: import 'github.com/nats-io/stan.go' is not allowed from list 'Main' (depguard)
	"github.com/nats-io/stan.go"
	^
eventbus/stan/sensor/trigger_conn.go:15:2: import 'github.com/nats-io/stan.go/pb' is not allowed from list 'Main' (depguard)
	"github.com/nats-io/stan.go/pb"
	^
eventsources/common/common.go:3:8: import 'github.com/cloudevents/sdk-go/v2/event' is not allowed from list 'Main' (depguard)
import "github.com/cloudevents/sdk-go/v2/event"
       ^
pkg/apis/events/event-data.go:23:2: import 'github.com/minio/minio-go/v7/pkg/notification' is not allowed from list 'Main' (depguard)
	"github.com/minio/minio-go/v7/pkg/notification"
	^
pkg/apis/events/event-data.go:24:2: import 'github.com/stripe/stripe-go' is not allowed from list 'Main' (depguard)
	"github.com/stripe/stripe-go"
	^
pkg/apis/events/event-data.go:26:2: import 'github.com/aws/aws-sdk-go/service/sqs' is not allowed from list 'Main' (depguard)
	sqslib "github.com/aws/aws-sdk-go/service/sqs"
	^
eventsources/eventing.go:14:2: import 'github.com/google/uuid' is not allowed from list 'Main' (depguard)
	"github.com/google/uuid"
	^
eventsources/eventing.go:18:2: import 'github.com/argoproj/argo-events/common/expr' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common/expr"
	^
eventsources/eventing.go:19:2: import 'github.com/argoproj/argo-events/common/leaderelection' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common/leaderelection"
	^
eventsources/eventing.go:20:2: import 'github.com/argoproj/argo-events/common/logging' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/common/logging"
	^
eventsources/eventing.go:21:2: import 'github.com/argoproj/argo-events/eventbus' is not allowed from list 'Main' (depguard)
	"github.com/argoproj/argo-events/eventbus"
	^
eventsources/eventing.go:23:2: import 'github.com/argoproj/argo-events/eventsources/common' is not allowed from list 'Main' (depguard)
	eventsourcecommon "github.com/argoproj/argo-events/eventsources/common"
```

**Comment by ryancurrah** [comment]
I think you should remove `depguard` from the enabled linters. We do not seem be using it, how I can tell is there is no list of allowed linters. Then you can resolve the rest of the lints easily.

**Comment by shyukri** [comment]
@ryancurrah @whynowy can you please re-check ?

File: Makefile [hunk 1]
```diff
@@ -38,6 +39,7 @@ override LDFLAGS += -X ${PACKAGE}.gitTag=${GIT_TAG}
 endif
 
 K3D ?= $(shell [ "`command -v kubectl`" != '' ] && [ "`command -v k3d`" != '' ] && [[ "`kubectl config current-context`" =~ k3d-* ]] && echo true || echo false)
+K3DCLUSTER?=e2e
```
[the remark this answers is not in the document]
Comment (goat-ssh) [reply Makefile]: done, and i've dropped the name in GHA

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -141,11 +141,11 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
       - name: Install k3d
-        run: curl -sfL https://raw.githubusercontent.com/rancher/k3d/main/install.sh | bash &
+        run: curl -s https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash &
       - name: Create k3d registry and cluster
         run: |
           k3d registry create e2e-registry --port 5111
-          k3d cluster create e2e -i rancher/k3s:v1.21.7-k3s1 --registry-use k3d-e2e-registry:5111
+          k3d cluster create -i rancher/k3s:v1.27.10-k3s2 --registry-use k3d-e2e-registry:5111
```
Comment (whynowy) [inline .github/workflows/ci.yaml:148]: Do not need to remove `e2e` here for the CI.

**Comment by blkperl** [comment]
@ryancurrah do you need any assistance?  I was considering opening a separate lint fix pr

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -141,11 +141,11 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
       - name: Install k3d
-        run: curl -sfL https://raw.githubusercontent.com/rancher/k3d/main/install.sh | bash &
+        run: curl -s https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash &
```
[the remark this answers is not in the document]
Comment (goat-ssh) [reply .github/workflows/ci.yaml]: This was the command fetched from k3d docs. Putting those back.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (shyukri) [reply Makefile]: Changed

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -141,11 +141,11 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
       - name: Install k3d
-        run: curl -sfL https://raw.githubusercontent.com/rancher/k3d/main/install.sh | bash &
+        run: curl -sfL https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash &
       - name: Create k3d registry and cluster
         run: |
           k3d registry create e2e-registry --port 5111
-          k3d cluster create e2e -i rancher/k3s:v1.21.7-k3s1 --registry-use k3d-e2e-registry:5111
+          k3d cluster create -i rancher/k3s:v1.27.10-k3s2 --registry-use k3d-e2e-registry:5111
```
Comment (ryancurrah) [inline .github/workflows/ci.yaml:148]:
```suggestion
          k3d cluster create e2e-i rancher/k3s:v1.27.10-k3s2 --registry-use k3d-e2e-registry:5111
```

File: Dockerfile [hunk 5]
```diff
@@ -2,13 +2,13 @@ ARG ARCH=$TARGETARCH
...
 ####################################################################################################
-FROM alpine:3.16.2 as base
+FROM alpine:3.19.1 as base
 ARG ARCH
 RUN apk update && apk upgrade && \
     apk add ca-certificates && \
     apk --no-cache add tzdata
 
-ENV ARGO_VERSION=v3.4.8
+ENV ARGO_VERSION=v3.5.4
```
Comment (ryancurrah) [inline Dockerfile:11]:
```suggestion
ENV ARGO_VERSION=v3.5.5
```

File: Makefile [hunk 6]
```diff
@@ -155,7 +157,7 @@ start: image
 	kubectl -n argo-events wait --for=condition=Ready --timeout 60s pod --all
 
 $(GOPATH)/bin/golangci-lint:
-	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b `go env GOPATH`/bin v1.52.1
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b `go env GOPATH`/bin v1.55.2
```
Comment (ryancurrah) [inline Makefile:160]:
```suggestion
	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b `go env GOPATH`/bin v1.57.2
```

File: third_party/prometheus-nats-exporter-docker/amd64/Dockerfile [hunk 7]
```diff
@@ -1,5 +1,5 @@
 # Golang binary building stage
-FROM golang:1.20.2
+FROM golang:1.21.6
```
Comment (ryancurrah) [inline third_party/prometheus-nats-exporter-docker/amd64/Dockerfile:2]:
```suggestion
FROM golang:1.21
```

**Review by ryancurrah — CHANGES_REQUESTED** [review]

File: third_party/prometheus-nats-exporter-docker/amd64/Dockerfile (hunk 7 above)
Comment (shyukri) [reply third_party/prometheus-nats-exporter-docker/amd64/Dockerfile:2]: It's hard to track down possible issues if the image is not static.

Comment (shyukri) [reply third_party/prometheus-nats-exporter-docker/amd64/Dockerfile]: Done

File: Makefile (hunk 6 above)
Comment (shyukri) [reply Makefile]: Done

File: Dockerfile (hunk 5 above)
Comment (shyukri) [reply Dockerfile]: Done

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (shyukri) [reply .github/workflows/ci.yaml:148]: Done

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (shyukri) [reply .github/workflows/ci.yaml:148]: Done

**Review by ryancurrah — APPROVED** [review]
LGTM

**Comment by yucombinator** [comment]
Hello! IS there any updates on this PR? We are blocked on our end pending the resolution of a few CVEs that are present in the older Go version used in the latest release.

**Comment by shyukri** [comment]
@whynowy any chance to take another look ? Or should we jump to golang 1.22 :) ?

**Comment by shyukri** [comment]
@whynowy re

**Pull request closed without merging**